pages/en/lb4/Testing-Your-Extensions.md

```diff
@@ -0,0 +1,305 @@
+- Prevent regressions when new features are added or bugs are fixed
+- Helps demonstrate usage
+
+## Project Setup
+
+If using `@loopback/cli` to create a new extension project, it will come with a
+`test` folder with a recommended folder for various types of tests.
+An automated test suite requires a test runner to execute the tests and we
+recommend `mocha`. It is enabled by default if `@loopback/cli` was used to
+create the extension project. This will install and configure `mocha` as well
+as create a `test` command in your `package.json`.
+
+Assertion libraries such as [ShouldJS](http://shouldjs.github.io/) (as `expect`),
+[SinonJS](http://sinonjs.org/), and a swagger validator are made available
+through the convenitent `@loopback/testlab` package. This is installed if the
+project was created using the CLI.
+
+### Manual Setup - Using Mocha
+- Install `mocha` by running `npm i mocha`. This will save the package in
+`package.json` as well.
+- In `package.json` add under `scripts` the following:
+`test: npm run build && mocha --recursive ./dist/test`
+
+## Types of tests
+
+A comprehensive test suite tests many aspects of your code. It is encouraged to
+test as many aspects of your application from a variety of perspectives to
+ensure correctness, integration and ensuring future compatibility. To accomplish
+this you should write unit, integration and acceptance tests.
+
+You may use any development methodology you want to write your extension, the
+important thing is to have an automated test suite.
+- **Traditional development:** write the code first and then write the tests
+- **Test-driven development:** write tests first, see them fail, then write the
+code to pass the tests
+
+### Unit Tests
+
+A unit test checks the smallest unit of code possible, in this case a function.
+This helps to ensure variable and state changes by outside actors don't affect
+the result. If the function has dependencies, you should substitute them with
+[test doubles](https://en.wikipedia.org/wiki/Test_double) to ensure only the
+function you are testing affects the results. SinonJS is our recommended
+framework to create test doubles in the form of spies, stubs, and mocks.
+
+#### Controllers
+
+At it's core, a controller is a simple class that is responsible for related
+actions on a object. Unit testing a controller in an extension is the same as
+unit testing a controller for an application.
+
+To test a controller you want to instantiate a new instance of your controller
+class and test a function, providing a test double for constructor arguments as
+needed.
+
+**`src/controllers/ping.controller.ts`**
+```ts
+export class PingController {
+  @get('/ping')
+  ping(msg?: string) {
+    return `You pinged with ${msg}`;
+  }
+}
+```
+
+**`test/unit/controllers/ping.controller.ts`**
+```ts
+import {PingController} from '../../..';
+import {expect} from '@loopback/testlab';
+
+describe('PingController() unit', () => {
+  it('.ping() (no arg)', () => {
```

Comment:
I thought our test practice was to use descriptive names for the test labels.
Something like `it('pings with no input', ...)`